File: python/src/smooai_observability/traceparent.py
"""W3C ``traceparent`` parse / format.

Python port of ``packages/core/src/traceparent.ts``, pinned by
``parity/sampling-corpus.json``::

    00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
    ^  ^                                ^                ^
    |  trace-id (16 bytes, 32 lowercase hex)             |
    |                                   span-id (8 bytes, 16 lowercase hex)
    version (1 byte, 2 hex)                              trace-flags (1 byte, 2 hex)

Parsing is STRICT — exactly four dash-separated fields, version exactly ``00``.
Rejected: wrong field count, wrong version (including the forbidden ``ff``),
non-hex or wrong-length fields, uppercase hex, an all-zero trace id, an all-zero
span id. The all-zero ids are the classic "propagated a placeholder" bug:
accepting them produces traces that all collide on ``000…0``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

_VERSION = "00"
_HEX32 = re.compile(r"^[0-9a-f]{32}$")
_HEX16 = re.compile(r"^[0-9a-f]{16}$")
_HEX2 = re.compile(r"^[0-9a-f]{2}$")


@dataclass(frozen=True)
class TraceContext:
    """A parsed ``traceparent``."""

    #: 32 lowercase hex chars.
    trace_id: str
    #: 16 lowercase hex chars.
    span_id: str
    #: trace-flags byte, 0-255.
    flags: int
    #: bit 0 of ``flags`` — the upstream sampling decision, which logs inherit.
    sampled: bool


def parse_traceparent(header: str) -> TraceContext | None:
    """Parse a ``traceparent`` header. Returns ``None`` for anything invalid."""
    parts = header.split("-")
    if len(parts) != 4:
        return None
    version, trace_id, span_id, flags_hex = parts
    if version != _VERSION:
        return None
    if not _HEX32.match(trace_id) or trace_id == "0" * 32:
        return None
    if not _HEX16.match(span_id) or span_id == "0" * 16:
        return None
    if not _HEX2.match(flags_hex):
        return None
    flags = int(flags_hex, 16)
    return TraceContext(trace_id=trace_id, span_id=span_id, flags=flags, sampled=flags & 0x01 == 0x01)


def format_traceparent(trace_id: str, span_id: str, flags: int | None = None, sampled: bool | None = None) -> str | None:
    """Format a ``traceparent`` header.

    Pass ``flags`` explicitly, or let it be derived from ``sampled``. An
    out-of-byte-range ``flags`` is rejected rather than silently truncated.

    Returns ``None`` rather than emitting a header a spec-compliant peer would
    reject — a malformed traceparent breaks correlation downstream just as
    thoroughly as a missing one, but silently.
    """
    resolved = flags if flags is not None else (1 if sampled else 0)
    if not isinstance(resolved, int) or isinstance(resolved, bool) or not 0 <= resolved <= 255:
        return None
    header = f"{_VERSION}-{trace_id}-{span_id}-{resolved:02x}"
    # Round-trip through the parser so format can never emit what parse rejects.
    return header if parse_traceparent(header) else None

File: rust/observability/src/traceparent.rs
//! W3C `traceparent` parse / format. Rust port of
//! `packages/core/src/traceparent.ts`, pinned by `parity/sampling-corpus.json`.
//!
//! ```text
//! 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
//! ^  ^                                ^                ^
//! |  trace-id (16 bytes, 32 lowercase hex)             |
//! |                                   span-id (8 bytes, 16 lowercase hex)
//! version (1 byte, 2 hex)                              trace-flags (1 byte, 2 hex)
//! ```
//!
//! Parsing is STRICT — exactly four dash-separated fields, version exactly
//! `00`. Rejected: wrong field count, wrong version (including the forbidden
//! `ff`), non-hex or wrong-length fields, uppercase hex, an all-zero trace id,
//! an all-zero span id. The all-zero ids are the classic "propagated a
//! placeholder" bug: accepting them produces traces that all collide on `000…0`.

const VERSION: &str = "00";

/// A parsed `traceparent`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceContext {
    /// 32 lowercase hex chars.
    pub trace_id: String,
    /// 16 lowercase hex chars.
    pub span_id: String,
    /// trace-flags byte.
    pub flags: u8,
    /// bit 0 of `flags` — the upstream sampling decision, which logs inherit.
    pub sampled: bool,
}

fn is_lower_hex(s: &str, len: usize) -> bool {
    s.len() == len
        && s.bytes()
            .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
}

/// Parse a `traceparent` header. Returns `None` for anything invalid.
pub fn parse_traceparent(header: &str) -> Option<TraceContext> {
    let parts: Vec<&str> = header.split('-').collect();
    if parts.len() != 4 {
        return None;
    }
    let (version, trace_id, span_id, flags_hex) = (parts[0], parts[1], parts[2], parts[3]);
    if version != VERSION {
        return None;
    }
    if !is_lower_hex(trace_id, 32) || trace_id.bytes().all(|b| b == b'0') {
        return None;
    }
    if !is_lower_hex(span_id, 16) || span_id.bytes().all(|b| b == b'0') {
        return None;
    }
    if !is_lower_hex(flags_hex, 2) {
        return None;
    }
    let flags = u8::from_str_radix(flags_hex, 16).ok()?;
    Some(TraceContext {
        trace_id: trace_id.to_string(),
        span_id: span_id.to_string(),
        flags,
        sampled: flags & 0x01 == 0x01,
    })
}

/// Format a `traceparent` header.
///
/// `flags` is an `i64` rather than a `u8` so an out-of-byte-range value from a
/// caller (or from the parity corpus, which pins 256) is *rejected* rather than
/// silently truncated. Pass `None` to derive the flags byte from `sampled`.
///
/// Returns `None` rather than emitting a header a spec-compliant peer would
/// reject — a malformed traceparent breaks correlation downstream just as
/// thoroughly as a missing one, but silently.
pub fn format_traceparent(
    trace_id: &str,
    span_id: &str,
    flags: Option<i64>,
    sampled: Option<bool>,
) -> Option<String> {
    let flags = flags.unwrap_or(if sampled.unwrap_or(false) { 1 } else { 0 });
    if !(0..=255).contains(&flags) {
        return None;
    }
    let header = format!("{VERSION}-{trace_id}-{span_id}-{flags:02x}");
    // Round-trip through the parser so format can never emit what parse rejects.
    parse_traceparent(&header).map(|_| header)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips() {
        let h = "00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01";
        let ctx = parse_traceparent(h).expect("valid");
        assert!(ctx.sampled);
        assert_eq!(
            format_traceparent(
                &ctx.trace_id,
                &ctx.span_id,
                Some(i64::from(ctx.flags)),
                None
            )
            .as_deref(),
            Some(h)
        );
    }

    #[test]
    fn format_never_emits_what_parse_rejects() {
        assert_eq!(
            format_traceparent("0".repeat(32).as_str(), "00f067aa0ba902b7", Some(1), None),
            None
        );
        assert_eq!(
            format_traceparent(
                "4bf92f3577b34da6a3ce929d0e0e4736",
                "00f067aa0ba902b7",
                Some(256),
                None
            ),
            None
        );
    }
}
